bin/xbps-rindex/index-add.c

```diff
@@ -281,86 +320,342 @@ index_add_pkg(struct xbps_handle *xhp, xbps_dictionary_t index, xbps_dictionary_
+			const char *pkgver = NULL, *arch = NULL;
+			xbps_dictionary_get_cstring_nocopy(pkg, "pkgver", &pkgver);
+			xbps_dictionary_get_cstring_nocopy(pkg, "architecture", &arch);
+			printf("stage: added `%s' (%s)\n", pkgver, arch);
+		}
+
+		xbps_object_iterator_release(iter);
+	}
+
+	return 0;
+}
+
+static int
+unstage(struct repo_state *state)
+{
+	xbps_object_iterator_t iter;
+	xbps_object_t keysym;
+
+	iter = xbps_dictionary_iterator(state->stage);
+	if (!iter)
+		return xbps_error_oom();
+
+	while ((keysym = xbps_object_iterator_next(iter))) {
+		const char *pkgver = NULL, *arch = NULL;
+		const char *pkgname = xbps_dictionary_keysym_cstring_nocopy(keysym);
+		xbps_dictionary_t pkg = xbps_dictionary_get_keysym(state->stage, keysym);
+
+		if (!xbps_dictionary_get_cstring_nocopy(pkg, "pkgver", &pkgver))
+			abort();
+		if (!xbps_dictionary_get_cstring_nocopy(pkg, "architecture", &arch))
+			abort();
+
+		printf("index: added `%s' (%s).\n", pkgver, arch);
+
+		if (!xbps_dictionary_set(state->index, pkgname, pkg)) {
+			xbps_object_iterator_release(iter);
+			return xbps_error_oom();
+		}
+	}
+	xbps_object_iterator_release(iter);
 
-	if (repo) {
-		index = xbps_dictionary_copy_mutable(repo->index);
-		stage = xbps_dictionary_copy_mutable(repo->stage);
-		meta = xbps_dictionary_copy_mutable(repo->idxmeta);
+	xbps_object_release(state->stage);
+	state->stage = NULL;
+
+	return 0;
+}
+
+static int
+commit(struct repo_state *states, unsigned nstates)
+{
+	struct shared_state shared;
+	int r;
+
+	shared.staged_shlibs = xbps_dictionary_create();
+	if (!shared.staged_shlibs)
+		return xbps_error_oom();
+
+	shared.used_shlibs = xbps_dictionary_create();
+	if (!shared.used_shlibs) {
+		xbps_object_release(shared.staged_shlibs);
+		return xbps_error_oom();
+	}
+
+	// collect all the used shared libraries in the stage
+	for (unsigned i = 0; i < nstates; i++) {
+		r = add_staged_shlib_providers(&shared, &states[i]);
+		if (r < 0)
+			return r;
+		r = find_used_staged_shlibs(&shared, &states[i]);
+		if (r < 0)
+			return r;
+	}
+
+	// throw out shared libraries that are already satisfied
+	for (unsigned i = 0; i < nstates; i++) {
+		r = purge_satisfied_by_index(&shared, &states[i]);
+		if (r < 0)
+			return r;
+		r = purge_satisfied_by_stage(&shared, &states[i]);
+		if (r < 0)
+			return r;
+	}
+
+	// ... now if there are libraries left, there is an inconsistency
+	if (xbps_dictionary_count(shared.used_shlibs) != 0) {
+		r = print_inconsistent_shlibs(&shared);
+		if (r < 0)
+			goto err;
+		r = print_staged_packages(states, nstates);
+		if (r < 0)
+			goto err;
 	} else {
-		index = xbps_dictionary_create();
-		stage = xbps_dictionary_create();
-		meta = NULL;
+		for (unsigned i = 0; i < nstates; i++) {
+			r = unstage(&states[i]);
+			if (r < 0)
+				goto err;
+		}
+	}
+
+	for (unsigned i = 0; i < nstates; i++) {
+		struct repo_state *state = &states[i];
+		// XXX: kinda useless to print for all repos, also print stage?.
```

Comment:
imo it would be useful to print the stage count too. maybe this could be prefixed by the repo path? 

```
/path/to/repo: index: %u packages registered.
```

Reply:
The comment about it being "useless" is that previously `xbps-rindex` could only add packages to one repo at a time and so printing the stats for just that repo that was being modified would've made sense. But theoretically with this change unless I'm missing something you could add all packages in different repositories with one xbps-rindex call.

So i guess it depends whether that is allowed or not.

Not sure if the distinction makes sense, since the xbps-rindex call that unstages the repo, unstages not just the repo we added to in that case printing stats for those repos too makes sense. Maybe its good to track and print stats only for modified repos.

Reply:
I think it would be ok (and more ergonomic and expected by the user) to allow adding packages to multiple repos in one invocation, especially when the repo can be modified during staging checks anyways

Reply:
```
Executing command [ xbps-rindex -v -R repo1 -R repo2 -a repo1/A-1.0_2.noarch.xbps ]
Fail: stdout not empty
--- /dev/null   2025-12-08 15:08:46.234592808 +0000
+++ /tmp/kyua.XduQkk/2/work/check.fkj0Ml/stdout 2025-12-09 20:37:00.678680807 +0000
@@ -0,0 +1,5 @@
+Inconsistent shlibs:
+  libruby.so.3.3 (provided by: ruby; used by: B)
+repo1: stage: added `A-1.0_2' (noarch)
+repo1: index: 5 packages registered.
+repo1: stage: 3 packages registered.
```

with the latest fixup, it will only print the newly added package and only print stats for repos that have been modified.